README.md
[![GitLab pipeline status](https://gitlab.com/zkovari/gradle-changelog-automation-plugin/badges/master/pipeline.svg)](https://gitlab.com/zkovari/gradle-changelog-automation-plugin/pipelines)
[![Maven central](https://img.shields.io/maven-central/v/org.zkovari.changelog/changelog-automation-gradle-plugin.svg)](https://search.maven.org/search?q=g:%20org.zkovari.changelog%20a:changelog-automation-gradle-plugin)
[![Documentation Status](https://readthedocs.org/projects/gradle-changelog-automation-plugin/badge/?version=latest)](https://gradle-changelog-automation-plugin.readthedocs.io/en/latest/?badge=latest)
[![codecov](https://codecov.io/gl/zkovari/gradle-changelog-automation-plugin/branch/master/graph/badge.svg)](https://codecov.io/gl/zkovari/gradle-changelog-automation-plugin)
[![Maintainability](https://api.codeclimate.com/v1/badges/4560682da831d5bdeb6f/maintainability)](https://codeclimate.com/github/zkovari/gradle-changelog-automation-plugin/maintainability)
[![Gitter](https://img.shields.io/gitter/room/zkovari/gradle-changelog-automation-plugin)](https://gitter.im/gradle-changelog-automation-plugin)
# Gradle Changelog Automation Plugin

Inspired by [GitLab](https://gitlab.com/gitlab-org/gitlab-ce/): [https://about.gitlab.com/2018/07/03/solving-gitlabs-changelog-conflict-crisis/](https://about.gitlab.com/2018/07/03/solving-gitlabs-changelog-conflict-crisis/)

## What does this plugin do

This plugin is able to generate `CHANGELOG.md` from YAML files.
One YAML file represents one unreleased changelog entry.
For example, a new added feature's entry would look like this:

```yaml
title: My new feature
reference: GH-1
author: John Doe
type: added
```

Then the plugin can process the unreleased YAML entries and combine them into `CHANGELOG.md`.
The changelog is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

As an example, see this project's changelog which is automatically generated by this plugin:
[CHANGELOG.md](CHANGELOG.md).

## Getting started

### How to apply the plugin

Using the [plugins DSL](https://docs.gradle.org/current/userguide/plugins.html#sec:plugins_block):

```gradle
plugins {
    id 'org.zkovari.changelog' version '0.4.0'
}
```

Or using legacy [plugin application](https://docs.gradle.org/current/userguide/plugins.html#sec:old_plugin_application):

```gradle
buildscript {
    repositories {
        jcenter()
    }
    dependencies {
        classpath 'org.zkovari.changelog:changelog-automation-gradle-plugin:0.4.0'
    }
}

apply plugin: 'org.zkovari.changelog'
```

### How to generate YAML entries

The plugin contains a generator script that you can fetch with the `fetchChangelogScript` task:

```bash
gradle fetchChangelogScript
# result: scripts/changelog.sh

# add permission
chmod +x scripts/changelog.sh
```

To generate a new unreleased changelog entry (YAML), run `changelog.sh`:

```bash
./scripts/changelog.sh --type added "My new feature"
```

As a result, a new changelog entry is generated under `changelogs/unreleased`,
in this case the following one:

```yaml
title: My new feature
reference:
author:
type: added
```

See `changelog.sh --help` or the [detailed documentation](https://gradle-changelog-automation-plugin.readthedocs.io/en/latest/2-user-guide.html#how-to-generate-unreleased-changelog-entries-yaml-files) for further information.

### How to process unreleased changelog entries

If you already have unreleased YAML entries under `changelogs/unreleased`, you can combine them into your `CHANGELOG.md` with the task `processChangelogEntries`:

```bash
gradle processChangelogEntries
```

As a result, your `CHANGELOG.md` is updated with a new release entry.
As an exmaple, see this project's changelog which is automatically generated by this plugin: [CHANGELOG.md](CHANGELOG.md).

The task `processChangelogEntries` also removes your unreleased entries from `changelogs/unreleased`.

## Further documentation

Please visit our [detailed documentation](https://gradle-changelog-automation-plugin.readthedocs.io/en/latest/) for further information.

## License

![GitHub](https://img.shields.io/github/license/zkovari/gradle-changelog-automation-plugin)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.4.0] - 2020-03-20
### Added
- Update Gradle to 6.2.2
- 36 Handle hashtag in references

## [0.3.0] - 2019-08-16
### Added
- 35 Accept shortened values for types in changelog.sh

### Changed
- 25 Do not display Help menu automatically in case of a wrong parameter

## [0.2.2] - 2019-07-21
### Fixed
- 15 Fix branch detection in changelog.sh and denote it in the generated filename

## [0.2.1] - 2019-07-20
### Changed
- Define project descriptions in the POM
- Define URL in the POM

## [0.2.0] - 2019-07-19
### Added
- Delete unreleased changelog entries after processing them into CHANGELOG.md

### Fixed
- Keep previous entries after updating changelog
- Close InputStream in changelog parsing

## [0.1.1] - 2019-07-18
### Fixed
- Handle empty String values in release entry generator
- Remove unnecessary quotes in generated YAML file

## [0.1.0] - 2019-07-18
### Added
- Initial implementation of changelog automation Gradle plugin


subprojects/changelog-automation-bin/src/main/resources/changelog.sh
#!/usr/bin/env bash
#
# Copyright 2019-2020 Zsolt Kovari
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# 
#-------------------------------------------------------------------------
# Authors:
#     Zsolt Kovari (zsolt@kovaridev.com)
#------------------------------------------------------------------------- 
# Script: changelog.sh
# Version: 0.3.0
# Last updated: 2020-03-18
# URL: https://github.com/zkovari/gradle-changelog-automation-plugin
# Issue tracker: https://github.com/zkovari/gradle-changelog-automation-plugin/issues
#-------------------------------------------------------------------------
# Usage: changelog.sh [OPTION]... --type [TYPE] [TITLE]
#        For the available options and examples, see changelog.sh --help 
#

# abort on error
set -e

# if you update it, update the header comment too as well as the date
VERSION=0.3.0

function display_help {
    cat <<Help
Usage: $(basename "$0") [OPTION]... --type [TYPE] [TITLE]

Generate a .yml file that corresponds to a new changelog entry.
The .yml file is generated under 'changelogs/unreleased' directory
relative to the current working directory where this script is executed.

Example: $(basename "$0") --type added "My new feature"

------------------------------------------------------------------------------

Obligatory:

  -t| --type [TYPE]     the type of the changelog entry. Available values:
                        added, changed, deprecated, fixed, removed, security.
                        The given value can be also shortened, e.g. '-t a'
                        corresponds to 'added', '-t secu' to 'security', etc.

Options for changelog content generation:

  -r| --reference [REF] write an extra reference to the changelog.
                        Typically reference of an issue or a merge/pull request
  -u| --git-username    write the current git user.name to the changelog

Miscellaneous:

  -h| --help            display this help text and exit
  -v| --version         display version information and exit
  --dry-run             display the .yml changelog entry without 
                        generating the file

------------------------------------------------------------------------------

Examples:

# 'added' changelog entry
$(basename "$0") --type added "Add new feature"

# 'removed' changelog entry
$(basename "$0") -t removed "Removed legacy behaviour"

# 'added' changelog entry with extra reference and git user name.
# Reference '18' could point to an issue or a merge/pull request
$(basename "$0") -t added -r 18 -u "Add new feature"

------------------------------------------------------------------------------

Author: Zsolt Kovari
Reference: https://github.com/zkovari/gradle-changelog-automation-plugin

Help
}

function checkEmptyArg {
    if [[ -z $1 ]]; then
        echo "$2 must be specified."
        echo "See options and examples in --help"
        exit 1
    fi
}

function getType {
    if [[ "added" =~ ^$1 ]]; then
        echo "added"
    elif [[ "changed" =~ ^$1 ]]; then
        echo "changed"
    elif [[ "deprecated" =~ ^$1 ]]; then
        echo "deprecated"
    elif [[ "fixed" =~ ^$1 ]]; then
        echo "fixed"
    elif [[ "removed" =~ ^$1 ]]; then
        echo "removed"
    elif [[ "security" =~ ^$1 ]]; then
        echo "security"
    else
        echo ""
    fi
}

function checkType {
	if [[ -z $(getType "$1") ]]; then
        echo "Invalid value was specified for --type: '$1'. Accepted values are: added, changed, deprecated, fixed, removed, security"
        echo "See options and examples in --help"
        exit 1
    fi		
}

OUTPUT_DIR="changelogs/unreleased"
TITLE=""
REFERENCE=""
AUTHOR=""
TYPE=""
USE_GIT_USERNAME=false
DRY_RUN=false

while [[ $# -gt 0 ]]
do
    key="${1}"
    case ${key} in
    -t|--type)
        TYPE="${2}"
        shift # past argument
        shift # past value
        ;;
    -r|--reference)
        REFERENCE="${2}"
        shift # past argument
        shift # past value
        ;;
    -u|--git-username)
        USE_GIT_USERNAME=true
        shift # past argument
        ;;
    -v| --version)
        echo "Version: $VERSION"
        exit 0
        shift # past argument
        ;;
    --dry-run)
        DRY_RUN=true
        shift # past argument
        ;;
    -h|--help)
        display_help
        exit 0
        shift # past argument
        ;;
    *)    # unknown option
        TITLE=$1        
        shift # past argument
        ;;
    esac
done

checkEmptyArg "$TITLE" "Title"
checkEmptyArg "$TYPE" "Type"
checkType "$TYPE"

if hash git 2>/dev/null; then
    if [[ "$USE_GIT_USERNAME" == true ]]; then
        AUTHOR=$(git config --get user.name)
    fi
fi

PARSED_TYPE=$(getType "$TYPE")
REFERENCE=${REFERENCE//#/__HASHTAG__}
CHANGELOG_ENTRY="# Auto-generated by $(basename "$0") script. Version: $VERSION
---
title: $TITLE
reference: $REFERENCE
author: $AUTHOR
type: $PARSED_TYPE"

if [[ "$DRY_RUN" = true ]]; then
    echo "$CHANGELOG_ENTRY"
    exit 0
fi

if [[ ! -d $OUTPUT_DIR ]]; then
    echo "Output directory is created: $OUTPUT_DIR"
    mkdir -p $OUTPUT_DIR 
fi

if hash git 2>/dev/null; then
    # get current branch name
    if git rev-parse --abbrev-ref HEAD >/dev/null 2>/dev/null; then
        FILENAME=$(git rev-parse --abbrev-ref HEAD 2>&1)
    else
        echo "WARN: Not a git repository: $(pwd)"
        FILENAME=""
    fi
else
    FILENAME="$(date +"%Y-%m-%d_%H-%M-%S")_changelog_entry"
fi
if [[ $FILENAME == "" || $FILENAME =~ "fatal: not a git" ]]; then
    FILENAME="$(date +"%Y-%m-%d_%H-%M-%S")_changelog_entry" 
fi

# replace unwanted characters to _
FILENAME=${FILENAME//[^a-zA-Z0-9_-]/_}
FILENAME="$FILENAME.yml"

FILEPATH="$OUTPUT_DIR/$FILENAME"
if [[ -f "$FILEPATH" ]]; then
    FILENAME="$(date +"%Y-%m-%d_%H-%M-%S")_$FILENAME"
    FILEPATH="$OUTPUT_DIR/$FILENAME"
fi

echo "$CHANGELOG_ENTRY" > "$FILEPATH"
echo "New changelog was generated to: $FILEPATH"
